app/uk/gov/hmrc/helptosavefrontend/controllers/EmailController.scala

```diff
@@ -242,31 +242,32 @@ class EmailController @Inject() (val helpToSaveService:          HelpToSaveServi
     }.merge
   }
 
-  def confirmEmail(email: String): Action[AnyContent] = authorisedForHtsWithNINO { implicit request ⇒ implicit htsContext ⇒
+  def confirmEmail(email: String): Action[AnyContent] = authorisedForHtsWithInfo { implicit request ⇒ implicit htsContext ⇒
     val nino = htsContext.nino
+    val decryptedEmailEither = EitherT.fromEither[Future](decryptEmail(email))
 
-      def doUpdate(hTSSession: HTSSession)(ifSuccess: ⇒ Result) = {
+      def doUpdate(hTSSession: HTSSession)(ifSuccess: ⇒ Future[Result]): Future[Result] = {
         val result = for {
-          e ← EitherT.fromEither[Future](decryptEmail(email))
+          e ← decryptedEmailEither
           _ ← sessionCacheConnector.put(hTSSession.copy(confirmedEmail = Some(e)))
           _ ← helpToSaveService.storeConfirmedEmail(e)
         } yield ()
 
-        result.fold[Result](
+        result.fold(
           { e ⇒
             logger.warn(s"Could not write confirmed email: $e", nino)
-            internalServerError()
+            toFuture(internalServerError())
           }, { _ ⇒
             ifSuccess
           }
-        )
+        ).flatMap(identity)
```

Comment:
Is a `fold(..).flatMap(identity)` necessary here. You could write it as part of the for-comprehension above:
```scala
val result = for {
  e ← decryptedEmailEither
  _ ← sessionCacheConnector.put(hTSSession.copy(confirmedEmail = Some(e)))
  _ ← helpToSaveService.storeConfirmedEmail()
 r   ← EitherT.liftF(ifSuccess)
  } yield r
```